Cancel and release extend to remote units of work

fosterseth opened the pull request:
issue #54

`cancelRemote()` uses a goroutine that connects to the remote node and issues a `work cancel` command. This goroutine will not timeout, but will terminate after a successful cancel or a release.

`ReleaseUnit()` will make a one-time attempt to connect to remote and issue a `work release` command. Regardless if it is successful, it will go ahead and delete local files.

Cancel and Release are robust to when nodes are down. If either the control node or remote node go down before the cancellation takes place, it will cancel the work when the nodes come back online.

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -689,6 +689,31 @@ func (w *Workceptor) CancelUnit(unitID string) error {
 		return fmt.Errorf("unknown work unit %s", unitID)
 	}
 	w.activeUnitsLock.RUnlock()
+	remoteUnitID := unit.status.RemoteUnitID
```
We should get a lock on `unit.lock` here.  (This was my error, we should have already had a lock before manipulating the unit status at the bottom of the function.)

ghjm commented:
Conceptually, do we want this to be "you call CancelUnit and it either works or fails, and then it's your problem if you want to retry?"  Or do we want this to be "you submit a cancel request and the monitor process will keep trying to cancel remotely until it eventually succeeds?"

fosterseth commented on pkg/workceptor/workceptor.go:
```diff
@@ -689,6 +689,31 @@ func (w *Workceptor) CancelUnit(unitID string) error {
+			logger.Error("Read error reading from %s: %s\n", remoteNodeID, err)
+			return err
+		}
+		cancelledRegex := regexp.MustCompile(fmt.Sprintf("cancelled.*%s", remoteUnitID))
```
maybe this is not the proper way to check. We could monitor the remote work status.State instead

ghjm replied:
Yes ^^

fosterseth commented on pkg/workceptor/workceptor.go:
```diff
@@ -322,6 +323,9 @@ func (w *Workceptor) monitorRemoteStatus(ctx context.Context, cancel context.Can
 			logger.Error("Error saving local status file: %s\n", err)
 			return
 		}
+		if unit.status.State == WorkStateFailed {
```
this makes sure we stop monitoring if the cancel goes through on the remote node.

ghjm replied:
Shouldn't we also stop monitoring the cancel if the status is WorkStateSucceeded?

Also, there's an `IsComplete` helper function, so you can do `if IsComplete(unit.status.State)` which returns true if it's failed or succeeded, and is perhaps a little more future-proof if we ever add more work states.

fosterseth commented on pkg/workceptor/workceptor.go:
```diff
@@ -631,6 +636,11 @@ func (w *Workceptor) scanForUnits() {
 				}
 				w.activeUnits[fi.Name()] = unit
 				if si.Node != "" && si.Node != w.nc.NodeID() {
+					if si.LocalCancel {
```
on control node restart, this will make sure we restart the cancellation goroutine

fosterseth commented on pkg/workceptor/workceptor.go:
```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
 }
 
+func (w *Workceptor) attemptRemoteAction(remoteNodeID, remoteUnitID, remoteAction string, unit *workUnit) error {
+	timeout := time.After(300 * time.Second)
```
times out after 5 minutes. This might need to be a configurable parameter

fosterseth commented on pkg/workceptor/workceptor.go:
```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+		}
+	}
+
+   if remoteAction == "cancel" {
```
`attemptRemoteAction` is used both for release and cancel, in the case of cancel we wait around to see if the status actually changes.

fosterseth commented on pkg/workceptor/workceptor.go:
```diff
@@ -631,6 +636,11 @@ func (w *Workceptor) scanForUnits() {
+						remoteUnitID := si.RemoteUnitID
+						go w.attemptRemoteAction(remoteNodeID, remoteUnitID, "cancel", unit)
+					}
 					go w.monitorRemoteUnit(unit, fi.Name())
```
even if LocalCancel is true, we still want to monitor the remote node until we get a WorkStateFailed from the remote status, thus we run the monitor even if we've issued the cancel.

fosterseth commented on pkg/workceptor/workceptor.go:
```diff
@@ -631,6 +636,11 @@ func (w *Workceptor) scanForUnits() {
 				}
 				w.activeUnits[fi.Name()] = unit
 				if si.Node != "" && si.Node != w.nc.NodeID() {
+					if si.LocalCancel && si.State == WorkStateRunning {
```
on control node restart, this will make sure we restart the cancellation goroutine

fosterseth replied:
I realize we need a way to determine if a previous cancel attempt timed out, and not re-try a cancel in that case. Otherwise, every time we restart the control service it will respawn the go routines to cancel previously timed out cancellations.

ghjm replied:
What if it's WorkStatePending?  Then we don't launch the monitor, and maybe the job goes to WorkStateRunning a second later.  Maybe this should be `if si.LocalCancel && !IsComplete(si.State)`.

fosterseth replied:
changed it to `!IsComplete(si.State)`

ghjm commented:
What if the timeout logic was:

- On a `cancel` request, we keep retrying indefinitely, including after a restart
- On a `release` request, we only make a single attempt and just report back an error if it fails

The reason I'm suggesting this is that I think a `cancel` without a `release` is the end-user telling us "I do still care about this job and I actively want it cancelled."  And sometimes cancelling a job might be really important (I want to cancel my `rm -rf /` job before it destroys everything).  So I don't think we really have a warrant to say "sorry, user, your cancel didn't go through and we're just giving up on it."  If the user wants us to give up on it, they have a way to do that - they can do a `release`.

On a `release` request, the user is saying "I no longer care about this job."  For all we know, the user has decommissioned the remote Receptor node and we're never going to talk to it again.  In this case I think we should make a best-reasonable-effort attempt at cancelling, and then just give up if it doesn't work.  But maybe the semantics here should be that we don't release control back to the caller until the release goes through or times out, and if it times out, we return an error message saying "remote node unreachable: remote job not released" (or words to that effect).  I don't think we need a timeout here - if the remote Receptor node is unresponsive, not in the routing table, etc, we just return this error immediately.  The reason I think we should hold the client response is that once the local release goes through, there's no way to inform the client that the remote release failed.

Thoughts?

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
 
+func (w *Workceptor) attemptRemoteAction(remoteNodeID, remoteUnitID, remoteAction string, unit *workUnit) error {
+	timeout := time.After(300 * time.Second)
+	retryInterval := 500 * time.Millisecond
```
Perhaps instead of a fixed interval, this should be a backoff scheme, where we increase the retry interval on failures, to some maximum limit.

fosterseth replied:
I've now implemented a backoff scheme

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+			logger.Error("%s of remote work unit timed out", remoteAction)
+			return true
+		default:
+			time.Sleep(retryInterval)
```
Instead of doing a `time.Sleep` in the `default` section, it would be better to do a `case <- time.After(retryInterval)`

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+	retryInterval := 500 * time.Millisecond
+	shouldTimeout := func() bool {
+		select {
+		case <-timeout:
```
The global timeout might be better treated as a `Context` that a caller could pass in.   Then we just check `context.Done()` and it's the caller's business if they want to put a timeout on the context or not.

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+				return nil
+			}
+
+			if shouldTimeout() {
```
It's a bit weird that we're calling a function named `shouldTimeout` that is actually doing the timeout.  Maybe it should be called `doTimeout`, or just be unrolled into this loop and not be a function at all.  (Also I think most of this logic goes away if we use a Context instead.)

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
 	//TODO: Handle remote cancellation
 	w.scanForUnits()
 	w.activeUnitsLock.RLock()
+	defer w.activeUnitsLock.RUnlock()
```
Golang can't stack locks, so if you're holding this lock all the way to the end of the function, you can't call any _other_ function that tries to acquire the lock.  I guess the original reason I did this was for the cancel call, which is now moved to a goroutine.  Are we sure the `unit.status.Save` call is okay?

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+	if unit.status.RemoteUnitID != "" {
+		remoteNodeID := unit.status.Node
+		remoteUnitID := unit.status.RemoteUnitID
+		go w.attemptRemoteAction(remoteNodeID, remoteUnitID, "cancel", unit)
```
Should we set the local `unit.status.Detail` to "Cancelling" or "Locally Cancelled" or something like that here?

fosterseth committed:
Cancel and release extend to remote units of work

Cancel will start a goroutine that continuously attempts
to connect remote node and issue a work cancel command.

Release will do a one-time attempt to connect to remote
and issue a release command. It will then proceed to delete
all local files related to that work.

issue #54

fosterseth committed:
remove extra lines

fosterseth committed:
start local cancel if state not complete

fosterseth committed:
re add logger info line

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -679,44 +714,146 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
 }
 
+func (w *Workceptor) cancelRemote(remoteNodeID, remoteUnitID string, unit *workUnit) error {
+	retryInterval := 500 * time.Millisecond
```
Maybe set the initial `retryInterval` to `SuccessWorkSleep` which we already have a constant for?

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -679,44 +714,146 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+				conn.Close()
+				break
+			}
+			conn.Close()
```
We should read the response to the cancel command and report it if it is an error.  Errors should always be of the form "ERROR: <description" so it should be sufficient to just check if the string starts with "ERROR".

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -679,44 +714,146 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+			retryInterval = MaxWorkSleep
+		}
+	}
+	retryInterval = 500 * time.Millisecond
+	var state int
+	for {
+		// check that unit still exists, could have been released
+		unit.lock.RLock()
+		released := unit.released
+		unit.lock.RUnlock()
+		if released {
+			return nil
+		}
+		unit.lock.RLock()
+		state = unit.status.State
+		unit.lock.RUnlock()
+		if IsComplete(state) {
+			return nil
+		}
+		time.Sleep(retryInterval)
+	}
```
What does this code do?

fosterseth replied:
this code checks that we are getting a job state change on the remote node. But I realize I don't actually do anything with this information, so it is kind of pointless.

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -679,44 +714,146 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
 		return fmt.Errorf("unknown work unit %s", unitID)
 	}
-	w.activeUnitsLock.RUnlock()
 	if unit.worker != nil {
```
This might be a bug in the existing code.  Is it legal to access `unit.worker` like this while not holding the lock on the unit?

fosterseth replied:
my understanding is that no fields of `unit` are read or touched in the calling of `Cancel()`

ghjm replied:
👍

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -679,44 +714,146 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+	nodeID := unit.status.Node
+	remoteUnitID := unit.status.RemoteUnitID
+	released := unit.released
+	unit.lock.Unlock()
```
Probably better not to unlock and immediately re-lock like this, because what happens if some other process changes these values during the interval when we're not holding the lock?

fosterseth replied:
yeah I didn't want the lock / unlock to cross over the `if` statement. I'm not sure of a more graceful way to handle that.

ghjm replied:
I don't think there's a more graceful way than just unlocking it at the right time in both the if and the else clause.

ghjm replied:
Actually you could just leave it locked until after the  `unit.status.Save()` below.  The `cancelRemote` call will block until the lock is released, but that's okay, because it's called as a goroutine, so it's not deadlocking _our_ thread.  If the goroutine happens to get scheduled before we're done with the lock, that's okay, it will just block until we release it.

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -679,44 +714,146 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+		unit.lock.Unlock()
+	}
+	unitdir := path.Join(w.dataDir, unitID)
+	w.activeUnitsLock.RLock()
```
I don't see why we need to be holding the activeUnitsLock here?

fosterseth replied:
yeah I think was supposed to be `unit.lock.RLock() ` instead

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -679,44 +714,146 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+			if err != nil {
+				logger.Error("Write error sending to %s: %s\n", nodeID, err)
+			}
+			conn.Close()
```
Would be nice to check the result here as well, and log an error ... although on a release, I guess we don't care as much.

fosterseth replied:
I now attempt to read the result here and report back if the remote node did not release correctly

ghjm commented:
I experimented with this and tried running it with the remote node up, down, etc.  Everything worked as I would expect.

One thing I found a bit disconcerting was that if you cancel a remote unit of work while the remote node is unreachable, you still get a response like `{"cancelled":"JvnNj9kH"}`.  I feel like if we haven't actually succeeded in doing the remote cancel, this response should say something like "cancel pending" ... maybe?  What do you think?

fosterseth committed:
used sync.waitgroup instead of bool; display 'cancel pending' if cancelling and waiting on remote

ghjm commented on pkg/workceptor/workceptor.go:
```diff
@@ -679,44 +714,146 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
+	w.activeUnitsLock.RUnlock()
+	if err != nil {
+		logger.Error("Error saving local status file: %s\n", err)
+		return nil
```
Shouldn't we return the actual error here?

fosterseth replied:
fixed this

fosterseth committed:
if cancel immediately succeeds, show the appropriate message

fosterseth committed:
cleaned up cancelRemote send to bool chan. Now logs errors when attempting to release remote units

fosterseth committed:
fixed linter

fosterseth committed:
make format changes

fosterseth committed:
add warning log
